From 1ea2e6d98ce36e07c4801d90c204c320ee2a58c4 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Sat, 16 Jan 2016 10:30:56 +0000
Subject: [PATCH 296/296] TV: Ensure OMX decoder releases shared buffers
 before videoout discards them

When AvFormatDecoder detects a change in size or frame rate in a MPEG or H264
clip it calls MythPlayer::SetVideoParams.  This in turn calls ReinitVideo which
can reallocate the video buffers.  The OMX decoder shares these buffers so it
is necessary to call PrivateDecoder::Reset before SetVideoParams.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythtv/libs/libmythtv/avformatdecoder.cpp    |    9 ++++++---
 mythtv/libs/libmythtv/privatedecoder_omx.cpp |    3 +--
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/mythtv/libs/libmythtv/avformatdecoder.cpp b/mythtv/libs/libmythtv/avformatdecoder.cpp
index 8568978..b0a2b82 100644
--- a/mythtv/libs/libmythtv/avformatdecoder.cpp
+++ b/mythtv/libs/libmythtv/avformatdecoder.cpp
@@ -3367,15 +3367,15 @@ void AvFormatDecoder::MpegPreProcessPkt(AVStream *stream, AVPacket *pkt)
 
             if (changed)
             {
+                if (private_dec)
+                    private_dec->Reset();
+
                 m_parent->SetVideoParams(width, height, seqFPS, current_aspect, kScan_Detect);
 
                 current_width  = width;
                 current_height = height;
                 fps            = seqFPS;
 
-                if (private_dec)
-                    private_dec->Reset();
-
                 gopset = false;
                 prevgoppos = 0;
                 firstvpts = lastapts = lastvpts = lastccptsu = 0;
@@ -3474,6 +3474,9 @@ int AvFormatDecoder::H264PreProcessPkt(AVStream *stream, AVPacket *pkt)
 
         if (fps_changed || res_changed)
         {
+            if (private_dec)
+                private_dec->Reset();
+
             m_parent->SetVideoParams(width, height, seqFPS, current_aspect, kScan_Detect);
 
             current_width  = width;
diff --git a/mythtv/libs/libmythtv/privatedecoder_omx.cpp b/mythtv/libs/libmythtv/privatedecoder_omx.cpp
index 8914a52..5190e10 100644
--- a/mythtv/libs/libmythtv/privatedecoder_omx.cpp
+++ b/mythtv/libs/libmythtv/privatedecoder_omx.cpp
@@ -567,9 +567,8 @@ OMX_ERRORTYPE PrivateDecoderOMX::FreeOutputBuffersCB()
         m_lock.unlock();
 
         VideoFrame *frame = HDR2FRAME(hdr);
-        if (frame)
+        if (frame && FRAME2HDR(frame) == hdr)
         {
-            assert(FRAME2HDR(frame) == hdr);
             FRAMESETHDR(frame, 0);
 
             AVBufferRef *ref = FRAME2REF(frame);
-- 
1.7.9.5

